ext/include/opentelemetry/ext/zpages/tracez_data_aggregator.h

```diff
@@ -0,0 +1,207 @@
+#pragma once
+
+#include <array>
+#include <atomic>
+#include <condition_variable>
+#include <list>
+#include <map>
+#include <mutex>
+#include <string>
+#include <thread>
+#include <unordered_set>
+#include <vector>
+
+#include "opentelemetry/ext/zpages/latency_boundaries.h"
+#include "opentelemetry/ext/zpages/tracez_processor.h"
+#include "opentelemetry/nostd/string_view.h"
+#include "opentelemetry/sdk/trace/span_data.h"
+#include "opentelemetry/trace/canonical_code.h"
+
+using opentelemetry::sdk::trace::SpanData;
+using opentelemetry::trace::CanonicalCode;
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace ext {
+namespace zpages {
+
+/**
+ * kMaxNumberOfSampleSpans is the maximum number of running, completed or error
+ * sample spans stored at any given time for a given span name.
+ * This limit is introduced to reduce memory usage by trimming sample spans
+ * stored.
+ */
+const int kMaxNumberOfSampleSpans = 5;
+
+/**
+ * TracezData is the span data to be displayed for tracez zpages that is
+ * stored for each span name.
+ */
+struct TracezData {
+  /**
+   * TODO: At this time the maximum count is unknown but a larger data type
+   * might have to be used in the future to store these counts to avoid overflow
+   */
+  unsigned int running_span_count;
+  unsigned int error_span_count;
+
+  /**
+   * completed_span_count_per_latency_bucket is an array that stores the count
+   * of spans for each of the 9 latency buckets.
+   */
+  std::array<unsigned int, kLatencyBoundaries.size()>
+      completed_span_count_per_latency_bucket;
+
+  /**
+   * sample_latency_spans is an array of lists, each index of the array
+   * corresponds to a latency boundary(of which there are 9).
+   * The list in each index stores the sample spans for that latency boundary.
+   */
+  std::array<std::list<std::shared_ptr<SpanData>>, kLatencyBoundaries.size()>
+      sample_latency_spans;
+
+  /**
+   * sample_error_spans is a list that stores the error samples for a span name.
+   */
+  std::list<std::shared_ptr<SpanData>> sample_error_spans;
+
+  /**
+   * sample_running_spans is a list that stores the running span samples for a
+   * span name.
```

Comment:
Curious - The spec is saying that span name could be updated, which means a span could start with name "foo", and end with name "bar". How do we plan to handle this (from experience perspective)?

Reply:
https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/api.md#span

Reply:
Oh, I wasn't aware of this. Can the span name change once it has ended as well? I am guessing not. One way is to go through the entire aggregated data and clear all running span information before starting the aggregation.